File: dsa/02-blind-75/arrays-hashing/01-two-sum/solutions/typescript/twoSumHashMap.ts
export function twoSum(numbers: number[], target: number): [number, number] {
  const indexByValue = new Map<number, number>();

  for (let index = 0; index < numbers.length; index += 1) {
    const value = numbers[index];
    const complement = target - value;
    const complementIndex = indexByValue.get(complement);

    if (complementIndex !== undefined) {
      return [complementIndex, index];
    }

    indexByValue.set(value, index);
  }

  throw new Error("No pair adds to the target");
}

File: dsa/02-blind-75/arrays-hashing/01-two-sum/solutions/python/two_sum_hash_map.py
def two_sum(numbers: list[int], target: int) -> list[int]:
    index_by_value: dict[int, int] = {}

    for index, value in enumerate(numbers):
        complement = target - value
        if complement in index_by_value:
            return [index_by_value[complement], index]
        index_by_value[value] = index

    raise ValueError("No pair adds to the target")
